src/testing/mock.rs

```diff
@@ -280,120 +170,90 @@ impl UnreliableRouter {
-            if rand_sample > self.reliability {
-                debug!("Simulated network fail.");
-                continue;
+            for peer_id in disconnected_peers.drain(..) {
+                self.peers.remove(&peer_id);
             }
-            let mut peers = self.peers.lock();
-            if let Some(peer) = peers.get_mut(&recipient) {
-                for hook in self.hook_list.lock().iter_mut() {
-                    hook.update_state(&mut data, sender, recipient);
+            for (sender, (mut data, recipient)) in buffer.drain(..) {
+                let rand_sample = rand::random::<f64>();
+                if rand_sample > self.reliability {
+                    debug!("Simulated network fail.");
+                    continue;
+                }
+                if let Some(peer) = self.peers.get_mut(&recipient) {
+                    for hook in self.hook_list.iter_mut() {
+                        hook.update_state(&mut data, sender, recipient);
+                    }
+                    peer.tx
+                        .unbounded_send((data, sender))
+                        .expect("channel should be open");
                 }
-                peer.tx
-                    .unbounded_send((data, sender))
-                    .expect("channel should be open");
             }
+            yield_now().await;
         }
-
-        Poll::Pending
     }
 }
 
-pub(crate) trait NetworkHook: Send {
-    fn update_state(&self, data: &mut NetworkData, sender: NodeIndex, recipient: NodeIndex);
-}
-
-#[derive(Clone)]
-pub(crate) struct AlertHook {
-    alerts_sent_by_connection: Arc<Mutex<HashMap<(NodeIndex, NodeIndex), usize>>>,
-}
-
-impl AlertHook {
-    pub(crate) fn new() -> Self {
-        AlertHook {
-            alerts_sent_by_connection: Arc::new(Mutex::new(HashMap::new())),
-        }
-    }
-
-    pub(crate) fn count(&self, sender: NodeIndex, recipient: NodeIndex) -> usize {
-        match self
-            .alerts_sent_by_connection
-            .lock()
-            .get(&(sender, recipient))
-        {
-            Some(count) => *count,
-            None => 0,
+async fn yield_now() {
+    let mut called = false;
+    poll_fn(move |ctx| {
+        if called {
+            Poll::Ready(())
+        } else {
+            called = true;
+            ctx.waker().wake_by_ref();
+            Poll::Pending
         }
-    }
+    })
+    .await
 }
 
-impl NetworkHook for AlertHook {
-    fn update_state(&self, data: &mut NetworkData, sender: NodeIndex, recipient: NodeIndex) {
-        use alerts::AlertMessage::*;
-        use network::NetworkDataInner::*;
-        if let crate::NetworkData(Alert(ForkAlert(_))) = data {
-            *self
-                .alerts_sent_by_connection
-                .lock()
-                .entry((sender, recipient))
-                .or_insert(0) += 1;
-        }
-    }
+pub trait NetworkHook: Send {
+    fn update_state(&mut self, data: &mut NetworkData, sender: NodeIndex, recipient: NodeIndex);
```

Comment:
I'm a bit unsure about the change from mutexes to mut refs – what was your reasoning here? I suspect it's likely correct, so the question is mostly curiosity.

Reply:
One obvious reason is that it is confusing when a method named update_state doesn't use a mutable reference. The other is that it forces you to use mutexes/other_ugliness even when Network owns some hook.

Reply:
The first argument is not particularly convincing, with Rust mutability abstractions it's expected that non-mut things change state if they have some runtime ownership thingies in them (like mutexes). But yeah, I guess if we can avoid mutexes that's convincing.